### ISSUE TITLE
Removing site does not lower tracked entities

### ISSUE THREAD

**Issue by toby12f** [narc0tiq/YARM#166]
I have been having UPS issues with YARM for a while now in my heavily modded save (1000h+) where it is causing on average 300ms lag per tick. I tried cutting the number of sites by more than half (it was very high), but it did nothing to improve UPS. I ran /c remote.call("YARM", "how_many_entities_tracked", game.player.name) before and after deleting sites but the number did not change at all. I tested this in a blank map with no other mods, but the number of entities tracked still did not change. Is it possible there is a problem with the site deletion process that results in "deleted" sites still being tracked?

EDIT: After going through the code it seems this is exactly what is happening. It deletes the "site" from the list of sites, but not the entities that were in the site. Whether a site is removed before or after depletion does not seem to make a difference. The list of entities scanned can only grow, it is never cleaned up. Was this by design?

**Comment by toby12f** [comment]
I modified the add_entity function in ore_tracker.lua to add entity.timestamp to each entity, passed from current_site.added_at. From there, I modified the remove_site function in resmon.lua to iterate through both the position_cache and the global.ore_tracker.entities table and sets all entities where the site.added_at for the site being removed matches the entity.timestamp. It also checks the ore type just as a failsafe. I also added global.ore_tracker.max since the length of ore_tracker cannot be used reliably as index anymore. After brief testing (like 5 minutes spamming stuff), it finally did not crash. I've never used Lua before and have no experience modding anything so it's very likely I broke a bunch of stuff elsewhere. I also know I have not accounted for migration from existing saves, but that can be dealt with later. This was mostly just a proof of concept for me.

`function resmon.on_click.remove_site(event)
    local site_name = string.sub(event.element.name, 1 + string.len("YARM_delete_site_"))

    local player = game.players[event.player_index]
    local force_data = global.force_data[player.force.name]
    local site = force_data.ore_sites[site_name]

    **if site.deleting_since then
        for k,v in pairs(global.ore_tracker.entities) do
            if v.timestamp == site.added_at and v.entity.name == site.ore_type then
                ore_tracker.position_cache[string.format("%d,%d", global.ore_tracker.entities[k].position.x * 100, global.ore_tracker.entities[k].position.y * 100)] = nil
                global.ore_tracker.entities[k] = nil
            end
        end**

        force_data.ore_sites[site_name] = nil

        if site.chart_tag and site.chart_tag.valid then
            site.chart_tag.destroy()
        end
    else
        site.deleting_since = event.tick
    end

    resmon.update_force_members_ui(player)
end`

and

`function ore_tracker.add_entity(entity, **added_at**)
    if not entity or not entity.valid or entity.type ~= "resource" then return nil end

    if not global.ore_tracker or not global.ore_tracker.entities then
        global.ore_tracker = {
            entities = {},
        }
        **global.ore_tracker.max = 1**
    end

    local position_key = position_to_string(entity.position)
    if ore_tracker.has_entity(entity) then
        local its_index = ore_tracker.position_cache[position_key]

        -- We're accessing the entity.position anyway, let's also use this
        -- opportunity to update the tracker values (and be 1000% certain
        -- that it's tracking the right entity).
        local tracking_data = global.ore_tracker.entities[its_index]
        tracking_data.entity = entity
        tracking_data.valid = entity.valid
        tracking_data.position = entity.position
        tracking_data.resource_amount = entity.amount

        return its_index
    end

    -- Otherwise, create the tracking data and store it, including position_cache
    local entities = global.ore_tracker.entities
    **local next_index = global.ore_tracker.max
    global.ore_tracker.max = global.ore_tracker.max + 1**

    entities[next_index] = {
        entity = entity,
        valid = entity.valid,
        position = entity.position,
        resource_amount = entity.amount,
        timestamp = added_at`

**Comment by toby12f** [comment]
I modified the add_entity function in ore_tracker.lua to add entity.timestamp to each entity, passed from current_site.added_at. From there, I modified the remove_site function in resmon.lua to iterate through both the position_cache and the global.ore_tracker.entities table and sets all entities where the site.added_at for the site being removed matches the entity.timestamp. It also checks the ore type just as a failsafe. I also added global.ore_tracker.max since the length of ore_tracker cannot be used reliably as index anymore. After brief testing (like 5 minutes spamming stuff), it finally did not crash. I've never used Lua before and have no experience modding anything so it's very likely I broke a bunch of stuff elsewhere. I also know I have not accounted for migration from existing saves, but that can be dealt with later. This was mostly just a proof of concept for me.

```
function resmon.on_click.remove_site(event)
    local site_name = string.sub(event.element.name, 1 + string.len("YARM_delete_site_"))

    local player = game.players[event.player_index]
    local force_data = global.force_data[player.force.name]
    local site = force_data.ore_sites[site_name]

    if site.deleting_since then
        for k,v in pairs(global.ore_tracker.entities) do
            if v.timestamp == site.added_at and v.entity.name == site.ore_type then
                ore_tracker.position_cache[string.format("%d,%d", global.ore_tracker.entities[k].position.x  100, global.ore_tracker.entities[k].position.y  100)] = nil
                global.ore_tracker.entities[k] = nil
            end
        end

        force_data.ore_sites[site_name] = nil

        if site.chart_tag and site.chart_tag.valid then
            site.chart_tag.destroy()
        end
    else
        site.deleting_since = event.tick
    end

    resmon.update_force_members_ui(player)
end

```

and

```
function ore_tracker.add_entity(entity, added_at)
    if not entity or not entity.valid or entity.type ~= "resource" then return nil end

    if not global.ore_tracker or not global.ore_tracker.entities then
        global.ore_tracker = {
            entities = {},
        }
        global.ore_tracker.max = 1
    end

    local position_key = position_to_string(entity.position)
    if ore_tracker.has_entity(entity) then
        local its_index = ore_tracker.position_cache[position_key]

        -- We're accessing the entity.position anyway, let's also use this
        -- opportunity to update the tracker values (and be 1000% certain
        -- that it's tracking the right entity).
        local tracking_data = global.ore_tracker.entities[its_index]
        tracking_data.entity = entity
        tracking_data.valid = entity.valid
        tracking_data.position = entity.position
        tracking_data.resource_amount = entity.amount

        return its_index
    end

    -- Otherwise, create the tracking data and store it, including position_cache
    local entities = global.ore_tracker.entities
    local next_index = global.ore_tracker.max
    global.ore_tracker.max = global.ore_tracker.max + 1

    entities[next_index] = {
        entity = entity,
        valid = entity.valid,
        position = entity.position,
        resource_amount = entity.amount,
        timestamp = added_at
```

**Comment by toby12f** [comment]
I have my implementation working quite reliably now, although I haven't tested it outside of basic use. I think I also fixed an issue where map pins were left behind occasionally. All that is left for me is to find a way to migrate data from previous versions. I can post it here once I figure it out. Fair warning, this is my first time using LUA or modding anything in general, so it's very likely there are better ways to achieve what I have done, and it's also likely there are a bunch of bugs I did not catch.

Separately, I noticed that extending sites was broken, although fixing that is beyond my abilities, and I have never had a use for it personally.
